인수테스트 기반 TDD 4단계 : 요금조회 PR 요청 드립니다. :)

ssonsh opened the pull request:
안녕하세요 동휘님 :) 벌써 이번 미션 마지막 단계 PR이네요!

3단계 미션에서 말씀주셨던 사항들 먼저 반영하고 4단계에서 요구하는 사항들에 대해 진행해보았습니다.
이번 과정에서 Inside Out, Outside In 과 같은 접근 방법에 대해 조금 더 명확하게 알게 되었고 이런 방법들을
ATDD와 TDD 로 접근하면서 조금씩 감을 잡아가고 있는 것 같다는 생각이 들면서도
여전히 어렵더군요 ㅎㅎ 

이번 단계 피드백도 잘 부탁드립니다. :)

ssonsh committed:
refactor (3차 피드백) : Exception 종류를 발생되는 케이스에 따라 명확한 분류로 정의

- asis : IllegalStateException
- tobe : IllegalArgumentException
- 요청된 인자가 잘못되어 발생된 Exception 임으로 IllegalArgumentException 을 발생시키도록 합니다.

ssonsh committed:
refactor : 적절하지 못한 enums 라는 패키지 분류를 재조정

- asis : enums.*
- tobe : domain.*
- 도메인 패키지 하위로 이동시킴으로써 명확한 분류 기준에 의해 위치하도록 변경

ssonsh committed:
refactor : 요청 시 전달된 토큰이 invalid 한 경우 AuthorizationException 을 throw 한다.

ssonsh committed:
docs(README) : 4단계 요금조회 기능 목록 1차 정의

ssonsh committed:
test(PathAcceptanceTest) : 지하철 경로 검색의 인수조건에 이용 요금이 함께 반환되는 조건을 추가하여 인수 테스트를 작성한다.

ssonsh committed:
feat(FareSectionType) : 이동 거리의 FareSectionType 구간에 따른 요금 정책 타입 정의, 정책 타입에 대한 테스트 코드 작성

- distance 거리에 따른 요금타입을 반환할 수 있다.

ssonsh committed:
refactor : FareSectionType 위치 변경

- in domain Package

ssonsh committed:
feat (FareCalculator) : 거리 기준에 따라 추가요금이 처리되어 요금을 계산해낼 수 있는 FareCalculator 구현

- FareSectionType 에 따라 요금이 계산되는 테스트 코드를 작성한다.

ssonsh committed:
feat (Fare) : Line에 추가 운임료인 Fare 를 추가한다.

- Fare 추가 운임료가 추가됨에 따라 작성된 테스트 코드의 변경사항을 대응한다.

ssonsh committed:
feat (Path, Line, Fare) : 출발역과 도착역까지 가는 구간에 있는 노선의 추가 운임료 중 가장 높은 추가 운임료를 확인할 수 있다.

- 노선에 적용된 구간 Section 에 대해 상->하, 혹은 하->상 으로의 이동을 확인할 수 있다.

- 노선이 1개 이상 (환승)된 경우 노선의 추가 운임료중 가장 높은 추가 운임료를 산정해낸다.

ssonsh committed:
feat (Path, Auth) : 노선 조회 시 로그인한 사용자만 노선을 조회할 수 있또록 한다.

- 로그인되지 않은 사용자가 노선을 조회하는 경우 UNAUTHORIZED Exception 이 발생한다.

- 로그인되어 있는 경우 토큰을 이용해 노선을 조회할 수 있으며, 정상적으로 노선이 조회된다.

ssonsh committed:
refactor (Path, Auth) : 비회원으로 경로 조회를 할 경우 Exception 발생이 아니라 정상적인 조회가 된다.

- 다만 연령별 요금할인 정책에 대한 부분이 적용되지 않을 뿐이다.

- 토큰이 없는 상태로 요청이 왔을 때 UnAuthorized Exception이 발생되던 부분 앞단에서 credential 정보의 유무에 따라 Empty LoginMember 객체를 반환시키도록 한다.

ssonsh committed:
docs(README) : 잘못 명시된 기능목록에 대한 기능목록 업데이트

ssonsh committed:
feat(DiscountAgeRuleType) : 나이에 따른 할인타입을 판단할 수 있는 enum 객체 생성

- 전달된 age 값을 통해 특정 구간의 type을 반환할 수 있다.

- 0~5 : TODDLER
- 6~12 : CHILDREN
- 13~18 : TEENAGER
- 19~MAX : ADULT

ssonsh committed:
refactor : enum values() 배열을 Arrays.stream() 으로 감싸기

ssonsh committed:
feat(DiscountAgeRuleType) : 연령별 구간에 따른 할인이 처리될 수 있다.

ssonsh committed:
test(PathAcceptanceTest) : 경로 조회 시 거리별, 노선별 요금 정책이 적용되며 추가적으로 로그인된 사용자가 경로 조회 시 사용자의 나이에 따라 연령별 할인 정책이 적용된다.

dhmin5693 commented on src/main/java/nextstep/subway/auth/application/AuthService.java:
```diff
@@ -29,10 +30,14 @@ public TokenResponse login(TokenRequest request) {
 
     public LoginMember findMemberByToken(String credentials) {
-        if (!jwtTokenProvider.validateToken(credentials)) {
+        if (Objects.isNull(credentials)) {
```
저는 `Objects.isNull`을 권장하지 않는 편입니다.
Object.isNull이나 == null 비교나 가독성 차이가 크지 않다고 생각하거든요.

이 메소드의 javadoc을 확인해볼까요?
```
API Note:
This method exists to be used as a java.util.function.Predicate, filter(Objects::isNull)
```
`Predicate` 를 위해서 만들었다고 쓰여있네요.

-----

위랑은 별개로, 여긴 보안상 허점이 될 가능성이 있습니다.
만약 회원만 사용 가능한 기능이 있다 해도 인증값을 null로 전달하면 무조건 비회원으로 넘어가게 됩니다.

`AuthenticationPrincipal` 에 required 와 같은 필드를 두어서 회원 전용 여부를 표기하면 어떨까요?

ssonsh replied:
Objects.isNull, Objects.nonNull 에 대한 부분이 그저 객체의 null 체킹용도로 사용될 수 있구나로 접근했는데
사실 사용의 용도가 Predicate 를 위해 만들어졌다는 것을 이번 피드백을 통해 알게되었네요.
실제로 어떤 역할을 하는 녀석인지 들여다 봤으면 충분히 알 수 있었는데 스스로가 조금 아쉽다고 느껴집니다 ㅎ

말씀주신대로 == null != null 과 가독성에 차이는 없다고 생각됩니다. :)

ssonsh replied:
API의 권한에 대해 AuthenticationPrincipal 이 필요하지만 required 옵션을 두어 반드시 토큰인증이 필요한지
혹은, 토큰인증은 필요없는지를 구분하도록 하였습니다.
- 하단 피드백주신 사항에 의해 LoginMember와 GuestMember 두가지 구현체로 구성하였는데 
- 인증 여부에 따라서 처리될 수 있도록 AuthenticationPrincialArgumentResolver의 resolveArgument 메소드를 변경하였습니다.

dhmin5693 commented on src/main/java/nextstep/subway/fare/calculator/FareCalculator.java:
```diff
@@ -0,0 +1,35 @@
+import nextstep.subway.line.domain.Sections;
+import nextstep.subway.path.domain.Path;
+
+public class FareCalculator {
```
static 메소드만 활용하는 클래스는 생성하지 못하도록 처리해주세요.

ssonsh replied:
한번더 꼼꼼히 봤으면 챙길 수 있었던 부분일 것 같은데 스스로가 아쉽습니다 ㅠ 
포인트 잘 잡아주셔서 감사합니다. :)

dhmin5693 commented on src/main/java/nextstep/subway/fare/calculator/FareCalculator.java:
```diff
@@ -0,0 +1,35 @@
+
+    private static final int ZERO_FARE = 0;
+
+    public static int calculate(Lines lines, Path path) {
```
`StringUtils.isEmpty()` 처럼 간단한 로직이 아니라 지금처럼 복잡한 로직을 서비스 메소드에서 의존하고 있으면 단위 테스트가 어려워지는 문제가 있어요.

요금은 그 자체로 도메인 객체화 할 수 있는 요소라고 생각이 듭니다.
연령별 요금은 인터페이스를 활용해서 다형성을 구현하면 enum을 통한 구현도 최소화할 수 있을 것 같아요.
리팩토링 해보시는걸 추천드립니다.

ssonsh replied:
연령별 요금은 인터페이스를 활용해서 다형성을 구현 << 이 포인트를 가지고 조금 더 고민하여 리팩토링을 진행해보겠습니다!

ssonsh replied:
DiscountPolicy 라는 인터페이스를 구성하고 
Non(할인없음), Free(공짜), Children(어린이), Teenager(청소년) 4가지의 할인정책을 구현해보았습니다.

구현된 DiscountPolicy는 ServiceMember(Guest or Login) 가 가질 수 있도록 구성해보았습니다. :)

그 과정속에서 자연스럽게 FareCalculator 라는 클래스 자체가 제거되었습니다!

구성의 방향성이 동휘님의 방향과 일치하는지는 다시한번 피드백을 통해 이뤄져야 할 것 같습니다. :)
개인적으론 이렇게 구성하고 진행해보니 과거 코드가 왜 저렇게 작성하였을까? 라는 생각이 들었습니다 ㅎ

dhmin5693 commented on src/main/java/nextstep/subway/line/domain/Fare.java:
```diff
@@ -0,0 +1,50 @@
+import javax.persistence.Embeddable;
+
+@Embeddable
+public class Fare {
```
이름에서 `추가 요금`의 느낌이 나면 더 좋지 않을까요?

ssonsh replied:
개인적으론 Line 노선의 Fare 라고 해서 노선의 요금이면 충분히 의미전달이 되겠구나 했지만
사실 추가 요금의 개념으로 명확히 명명하는게 더 나은 방향이라 생각됩니다! :)

dhmin5693 commented on src/main/java/nextstep/subway/path/application/PathService.java:
```diff
@@ -24,13 +26,23 @@ public PathService(LineService lineService, StationService stationService) {
+    private int calculatePathFare(LoginMember loginMember, Lines lines, Path path) {
+        int fare = FareCalculator.calculate(lines, path);
+
+        if (loginMember.isLogin()) {
```
`loginMember.isLogin()` 이라는 구문이 살짝 어색해보여요.
`LoginMember`는 로그인 한 사용자만 사용할 수 있을법한 이름이거든요.
현재 상황에선 비로그인 사용자도 `LoginMember`를 사용하고 있습니다.

ssonsh replied:
너무 단순하게 생각하고 작업한 내용이었다 싶습니다. 
조금 더 고민하여 로그인한 사용자와 그렇지 않은 사용자를 구분하는 개념을 고민해보겠습니다.
말그대로 LoginMember는 로그인한 회원이라는 의미를 가지고 있는데 이것 자체가 로그인하지 않은 사용자를 판단하는 것이 맞는지도 고민스럽군요!

ssonsh replied:
서비스를 사용하는 사용자라는 관점에서 ServiceMember 인터페이스를 구성하였고
이를 구현한 기존 로그인된 사용자를 의미하는 LoginMember와 로그인되지 않은 사용자를 의미하는 GuestMember 두가지로 구현체를 생성하였습니다. 

이후 Authentication 과정에서 인증이 필요한지 여부(required = true or false) 에 따라서 LoginMember를 반환하거나
GuestMember를 반환하는 처리로 변경해보았습니다. :)

dhmin5693 reviewed (changes requested):
미션 잘 수행해주셨네요 😄

다만 마지막 미션이니만큼 조금 더 생각해볼만한 여지를 남겨두었어요.
확인 후 재요청 부탁드립니다. 🙏

ssonsh committed:
refactor : Objects.isNull, Objects.nonNull 사용 부분에 대해 == null, != null 로 변경

- 처리하는 방향은 동일하지만 사용에 대해서 목적이 다르다는 부분을 알게 되었음.

- Objects.isNull, nonNull 은 Predicate를 위해 만들어진 녀석임.

- 가독성 적인 측면에도 차이가 없음.

ssonsh committed:
refactor : static 메소드만 활용하는 클래스는 별도로 객체를 생성하지 못하도록 private 생성자 추가

ssonsh committed:
refactor (Fare -> AdditionalFare) : 노선 (Line)에 대한 추가 요금 도메인을 Fare 로 정의하였으나 "추가요금" 이라는 의미를 좀더 명확히 표현하기 위해 리네임을 진행함.

ssonsh committed:
refactor : 서비스를 사용하는 사용자 관점에서 인증을 접근한다.

- 로그인된 사용자는 LoginMember
- 로그인되지 않은 사용자는 GuestMember 로 처리할 수 있도록 한다.

ssonsh committed:
refactor(AuthenticationPrincipal) : required 옵션을 추가하여 반드시 로그인되어야 사용할 수 있는 서비스와 그렇지 않아도 사용될 수 있는 서비스를 처리한다.

- required 값에 따라서 LoginMember와 GuestMember를 처리하는 AuthService 메소드를 제공한다.

- 관련 처리는 AuthenticationPrincipalArgumentResolver 에서 애노테이션의 required 값을 판단하여 처리하도록 구성한다.

ssonsh committed:
refactor (Fare) : 추가요금에 대한 도메인 생성

- 기존 FareCalculator 를 제거하며 이를 Fare 도메인이 대체한다.

- 추가요금, 노선요금에 대한 Calculate 처리를 Fare 도메인 내에서 처리할 수 있도록 리팩토링한다.

- 변경된 추가요금 Fare 에 대해 PathService 내에서의 요금 관련 처리 로직이 변경된다.

- 현재 형상은 연령별 요금할인은 처리되지 않으며, 관련 테스트 코드도 Failed 상태

ssonsh committed:
refactor (DiscountPolicy, ServiceMember, Fare) : 연령별 할인정책에 대한 로직 리팩토링

- 기존 Enum을 활용한 처리를 진행하였으나, 이 부분을 할인정책이라는 개념으로 추상화하고 각각의 할인 정책을 구현하는 방향으로 리팩토링을 진행

- ServiceMember는 기존 리팩토링을 통하여 Guest와 Login 두가지로 분리되었으며 각 할인정책을 적용시키도록 함.

- 할인정책은 Non(할인없음), Free(0원), Children(어린이), Teenager(청소년) 총 4가지의 정책을 구현함.

- 연령별 할인정책이 적용되어 최종 Fare 정보를 반환하는 로직에 대해 대응

dhmin5693 commented on src/main/java/nextstep/subway/fare/domain/Fare.java:
```diff
@@ -0,0 +1,46 @@
+
+public class Fare {
+
+    private static int ZERO_FARE = 0;
```
상수는 final 을 붙여주는게 좋지 않을까요?

ssonsh replied:
상수임에도 final을 안붙였네요 ㅠㅠ 좀더 꼼꼼히 챙기겠습니닷!

dhmin5693 commented on src/main/java/nextstep/subway/fare/domain/Fare.java:
```diff
@@ -0,0 +1,46 @@
+    private static int ZERO_FARE = 0;
+
+    private int overFare;
+    private int lineFare;
```
필드가 한번 생성되면 변화하지 않아서 final로 불변을 보장해줘도 좋을 것 같습니다.

ssonsh replied:
불변하는 필드에 대한 final 키워드 붙이기!! 이 부분도 의식적으로 접근하겠습니다 :)

dhmin5693 commented on src/main/java/nextstep/subway/auth/domain/GuestMember.java:
```diff
@@ -0,0 +1,26 @@
+
+    @Override
+    public DiscountPolicy getDiscountPolicy() {
+        return new NonDiscountPolicy();
```
`DiscountPolicy`를 상속하는 모든 인스턴스는 싱글턴으로 만들어두고 가져오면 성능상 이점을 가질 수 있겠네요.

ssonsh replied:
:) 이 부분은 미쳐 생각하지 못했는데 싱글턴으로 구성함으로써 성능상 이점을 가져갈 수 있음을 충분히 인지하였습니다 !

dhmin5693 commented on src/main/java/nextstep/subway/auth/domain/ServiceMember.java:
```diff
@@ -0,0 +1,10 @@
+
+import nextstep.subway.policy.DiscountPolicy;
+
+public interface ServiceMember {
```
다형성 구현 💯 
getter 상속이 조금 미묘하긴 한데, 큰 상관 없을 것도 같네요.

dhmin5693 commented on src/main/java/nextstep/subway/policy/DiscountPolicy.java:
```diff
@@ -0,0 +1,16 @@
+
+import nextstep.subway.fare.domain.FareType;
+
+public interface DiscountPolicy {
```
다형성 구현 💯
이렇게 구현하셨으면 하는 마음으로 피드백을 남겼는데, 잘 작성해주셨네요. 😄

ssonsh replied:
이렇게 구현하고 보니 다형성을 이렇게 활용하는 것이였는데 왜 미쳐 생각하지 못했을까! 라는 생각이 들었습니다 ㅎ

dhmin5693 commented on src/test/java/nextstep/subway/fare/domain/FareTest.java:
```diff
@@ -0,0 +1,151 @@
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+class FareTest {
```
`AdditionalFareTest` 가 더 적절하지 않을까요?

ssonsh replied:
Fare -> AdditionalFare 로 변경하는 과정에서 리네임을 챙기지 못하였던 부분인 것 같습니다 ㅠ

dhmin5693 reviewed (approved):
피드백 잘 반영해주셨네요 😄

요금 정책을 다형성으로 구현하셨으면 했는데 제 의도를 자세히 설명하지 않은 것 같아 걱정했었습니다.
다행히도 기우가 되었군요.

미션을 너무 잘 수행해주셔서 이걸로 마무리해도 될 것 같습니다.
더 이상 미션과는 관계없지만 약간 개선했으면 하는 부분에 코멘트를 남겨두었어요.
나중에 천천히 확인해주세요 😄

이걸로 미션의 절반 이상이 마무리되었는데요,
마지막까지 얼마 남지 않은 만큼 조금만 더 힘내시기 바랍니다. 💪
그 동안 수고 많으셨습니다.
